Support query params in asynchronously loaded routes

trentmwillis opened the pull request:
This PR enables support for query params in routes that are lazily loaded while not altering existing behavior. There are lots of details about the changes in the individual commit messages.

The bulk of the changes center on two key themes:
1. Not using the `handler` (route) before it is absolutely needed
2. Making it possible to decouple the Router's logic from the presence of Routes

rwjblue commented on packages/ember/tests/routing/query_params_test/query_param_async_get_handler_test.js:
```diff
@@ -0,0 +1,80 @@
+      location: 'test',
+      _getHandlerFunction() {
+        let getHandler = this._super(...arguments);
+        return (routeName) => Promise.resolve().then(() => getHandler(routeName));
```
Likely should use off of `RSVP.Promise` (since we support platforms without native Promise).  Also, when you make that change you should use "real" async and make a helper method to resolve in a `setTimeout` or something...

trentmwillis replied:
The problem with always deving in Chrome...will fix. I'll be sure to add a small setTimeout. Still planning to add several more tests.

rwjblue reviewed (changes requested):
Changes mostly look good, only one small nit-pick

trentmwillis commented:
@rwjblue made the requested changes. I have at least one more test to write and want to verify this works in the ember-engines repo, but the functional code is pretty much done. Also added a lot of docs around what is going on.

trentmwillis commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -653,22 +654,94 @@ const EmberRouter = EmberObject.extend(Evented, {
+      return JSON.stringify(value);
+    }
+
+    return `${value}`;
```
This was copied form the original function in route.js. That said, I actually like this as a way of coercing a string value.

homu commented:
:umbrella: The latest upstream changes (presumably #14475) made this pull request unmergeable. Please resolve the merge conflicts.

Commit:
Introduce _setRouteName to allow easy access to fullRouteName in Engines

Commit:
Don't call getHandler from _queryParamsFor

`Router#_queryParamsFor` now takes a `handlerInfos` array instead of a route name so that it doesn't need to call `getHandler` on the Router.js instance. This is important because an asynchronously loaded Route will return a `Promise` from `getHandler`.

To support the above, several functions now take either a `TransitionState` object or an array of `HandlerInfo` objects instead of a route name and contexts/models to calculate the state.

Commit:
Don't call HandlerInfo#becomeResolved

When calculating post-transition state, we no longer call
becomeResolved because it will attempt to fetch the handler
for that route even though we don't need it to serialize
params information.

Commit:
Add Router support for QP meta not being present

`_fullyScopeQueryParams` and `_hydrateUnsuppliedQueryParams` now have logic to skip over handlers that haven't loaded and thus can't give `qpMeta` information. Due to this, `_queryParamsFor` will not cache the result of an invocation where some `qpMeta` is missing. This way, when the `qpMeta` is available, it can then be accessed and cached. This solves the issue where QPs were never serializing into the URL during `finalizeQueryParamChange`.

Commit:
Move (de)serialization logic for QPs into Router

Commit:
Cache QP meta in BucketCache

This will allow lazily loaded routes/engines to use the BucketCache as its source of truth for QP meta information instead of having to go directly to the Route instance.

Commit:
Add tests for query params with asynchronously loaded routes

rwjblue commented on packages/ember-routing/lib/system/route.js:
```diff
@@ -142,14 +142,37 @@ let Route = EmberObject.extend(ActionHandler, Evented, {
+    @method setRouteName
+    @param {String} name
+  */
+  setRouteName(name) {
```
We should use a symbol or underscore any newly added private methods to make it super obvious that they are private (and avoid naming collisions with userland methods/properties).

rwjblue commented:
Awesome, thanks @trentmwillis!
